source/common/config/grpc_mux_subscription_impl.h

```diff
@@ -1,77 +1,86 @@
 #pragma once
 
-#include "envoy/config/ads.h"
+#include "envoy/config/grpc_mux.h"
 #include "envoy/config/subscription.h"
 
 #include "common/common/assert.h"
 #include "common/common/logger.h"
 #include "common/grpc/common.h"
 #include "common/protobuf/protobuf.h"
+#include "common/protobuf/utility.h"
 
 #include "api/discovery.pb.h"
 
 namespace Envoy {
 namespace Config {
 
 /**
- * Adapter from typed Subscription to untyped AdsApi. Also handles per-xDS API stats/logging.
+ * Adapter from typed Subscription to untyped GrpcMux. Also handles per-xDS API stats/logging.
  */
 template <class ResourceType>
-class AdsSubscriptionImpl : public Subscription<ResourceType>,
-                            AdsCallbacks,
-                            Logger::Loggable<Logger::Id::config> {
+class GrpcMuxSubscriptionImpl : public Subscription<ResourceType>,
+                                GrpcMuxCallbacks,
+                                Logger::Loggable<Logger::Id::config> {
 public:
-  AdsSubscriptionImpl(AdsApi& ads_api, SubscriptionStats stats)
-      : ads_api_(ads_api), stats_(stats),
+  GrpcMuxSubscriptionImpl(GrpcMux& grpc_mux, SubscriptionStats stats)
+      : grpc_mux_(grpc_mux), stats_(stats),
         type_url_(Grpc::Common::typeUrl(ResourceType().GetDescriptor()->full_name())) {}
 
   // Config::Subscription
   void start(const std::vector<std::string>& resources,
              SubscriptionCallbacks<ResourceType>& callbacks) override {
     callbacks_ = &callbacks;
-    watch_ = ads_api_.subscribe(type_url_, resources, *this);
+    watch_ = grpc_mux_.subscribe(type_url_, resources, *this);
     // The attempt stat here is maintained for the purposes of having consistency between ADS and
     // gRPC/filesystem/REST Subscriptions. Since ADS is push based and muxed, the notion of an
     // "attempt" for a given xDS API combined by ADS is not really that meaningful.
     stats_.update_attempt_.inc();
   }
 
   void updateResources(const std::vector<std::string>& resources) override {
-    watch_ = ads_api_.subscribe(type_url_, resources, *this);
+    watch_ = grpc_mux_.subscribe(type_url_, resources, *this);
     stats_.update_attempt_.inc();
   }
 
-  const std::string versionInfo() const override { NOT_IMPLEMENTED; }
+  const std::string versionInfo() const override { return version_info_; }
```

Comment:
As I understand it, the version here will be the last returned version from a discovery response, regardless of MUXd resource, right? I think that's fine but might be worth a comment.

Reply:
This will be for the specific type URL of the resource. There is one `GrpcMuxSubscriptionImpl` per type URL.

Reply:
Sorry where is that setup? (Just trying to understand)

Reply:
Ah I see: https://github.com/envoyproxy/envoy/pull/1658/files#diff-1adb93ed0d49e43e5ed44c13ff6ea6cbR78

OK got it. Makes sense. This is neat. It came together really well.